modules/selection/src/main/kotlin/com.r3.corda.lib.tokens.selection/api/Selector.kt

```diff
@@ -0,0 +1,274 @@
+     *
+     * Set [TokenQueryBy.issuer] to specify issuer.
+     * Calling selectTokens multiple time with the same lockId will return next unlocked states.
+     *
+     * @param externalId external id that states should be selected from, if null then states not mapped to any id will be chosen
+     * @param lockId id used to lock the states for spend, defaults to [FlowLogic] runID
+     * @param requiredAmount amount that should be spent
+     * @param queryBy narrows down tokens to spend, see [TokenQueryBy]
+     * @return List of [FungibleToken]s that satisfy the amount to spend, empty list if none found.
+     */
+    // From external id
+    @Suspendable
+    fun selectTokens(
+            externalId: UUID? = null,
+            lockId: UUID = FlowLogic.currentTopLevel?.runId?.uuid ?: UUID.randomUUID(),
+            requiredAmount: Amount<TokenType>,
+            queryBy: TokenQueryBy = TokenQueryBy()
+    ): List<StateAndRef<FungibleToken>> {
+        return selectTokens(Holder.fromUUID(externalId), lockId, requiredAmount, queryBy)
+    }
+
+    /**
+     * Select [FungibleToken]s that cover [requiredAmount] that belong to [holdingKey]. Notice that this
+     * function doesn't calculate change. If query criteria is not specified then only held token amounts are used.
+     *
+     * Set [TokenQueryBy.issuer] to specify issuer.
+     * Calling selectTokens multiple time with the same lockId will return next unlocked states.
+     *
+     * @param holdingKey key that holds the states to select
+     * @param lockId id used to lock the states for spend, defaults to [FlowLogic] runID
+     * @param requiredAmount amount that should be spent
+     * @param queryBy narrows down tokens to spend, see [TokenQueryBy]
+     * @return List of [FungibleToken]s that satisfy the amount to spend, empty list if none found.
+     */
+    // From a given key
+    @Suspendable
+    fun selectTokens(
+            holdingKey: PublicKey,
+            lockId: UUID = FlowLogic.currentTopLevel?.runId?.uuid ?: UUID.randomUUID(),
+            requiredAmount: Amount<TokenType>,
+            queryBy: TokenQueryBy = TokenQueryBy()
+    ): List<StateAndRef<FungibleToken>> {
+        return selectTokens(Holder.KeyIdentity(holdingKey), lockId, requiredAmount, queryBy)
+    }
+
+    /**
+     * TODO
+     */
+    // Token only
+    @Suspendable
+    fun generateMove(
+            lockId: UUID = FlowLogic.currentTopLevel?.runId?.uuid ?: UUID.randomUUID(),
+            partiesAndAmounts: List<Pair<AbstractParty, Amount<TokenType>>>,
+            changeHolder: AbstractParty,
+            queryBy: TokenQueryBy = TokenQueryBy()
+    ): Pair<List<StateAndRef<FungibleToken>>, List<FungibleToken>> {
+        return generateMove(Holder.TokenOnly, lockId, partiesAndAmounts, changeHolder, queryBy)
+    }
+
+    /**
+     * TODO
+     */
+    // External id
+    @Suspendable
+    fun generateMove(
+            externalId: UUID? = null,
```

Comment:
This is for cases when node has keys that are UNMAPPED to external id. You want to have ability to query for them too. But... VaultQueryCriteria doesn't take into account that logical case!
To answer your second comment, no, it won't be handled using queryCriteria, queryCriteria should be removed in 2.0 it's there only to avoid api breaking in minor release. We don't have translation from query criteria into in memory selection, you would have to support arbitrary queries.